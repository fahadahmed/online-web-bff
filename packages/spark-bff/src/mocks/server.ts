import { setupServer } from 'msw/node';
import a2pCustomerHandlers from 'datasources/a2p/customer/mocks/handlers';
import accessHandlers from 'datasources/user/access/mocks/handlers';
import accountSummaryHandlers from 'datasources/customer/accountsService/mocks/handlers';
import addressHandlers from 'datasources/location/address/mocks/handlers';
import airpointsHandlers from 'datasources/airpoints/mocks/handlers';
import availableNumbers from 'datasources/shop/availableNumbers/mocks/handlers';
import balanceServiceHandlers from 'datasources/finance/balanceService/mocks/handlers';
import billPaymentHandlers from 'datasources/finance/paymentService/billPayment/mocks/handlers';
import cartServicesHandlers from 'datasources/shop/cart/mocks/handlers';
import checkoutServicesHandlers from 'datasources/shop/checkout/mocks/handlers';
import connectionDiagnosticsHandlers from 'datasources/line/assure/connectionDiagnostics/mocks/handlers';
import connectionPromiseHandler from 'datasources/line/assure/connectionPromise/mocks/handlers';
import contentServiceHandlers from 'datasources/utility/content/mocks/handlers';
import digitalBillHandlers from 'datasources/finance/digitalBill/mocks/handlers';
import identityServiceHandlers from 'datasources/user/identityService/mocks/handlers';
import lineConfiguarationHandlers from 'datasources/line/lineConfiguration/mocks/handlers';
import lineDetailServiceHandlers from 'datasources/line/lineDetails/mocks/handlers';
import lineDetailsHandlers from 'datasources/line/lineSummary/mocks/handlers';
import lineSpendHistoryHandlers from 'datasources/line/spend/mocks/handlers';
import lineSubscriptionsService from 'datasources/line/lineSubscriptionsService/mocks/handlers';
import lineUsageDetailsMocks from 'datasources/line/usage/usageDetailsService/mocks';
import lineUsageHistoryMocks from 'datasources/line/usage/usageHistoryService/mocks';
import lineUsageMocks from 'datasources/line/usage/mocks/handlers';
import lineUsageTransactionMocks from 'datasources/line/usage/usageTransactionsService/mocks';
import networkSettingsHandlers from 'datasources/technology/networkSettingsService/mocks/handlers';
import orderHandlers from 'datasources/shop/order/mocks/handlers';
import paymentServiceHandlers from 'datasources/finance/paymentService/mocks/handlers';
import pointOfAddressHandlers from 'datasources/location/pointOfInterest/mocks/handlers';
import pointOfOutageHandlers from 'datasources/location/pointOfInterest/outagesPointOfInterest/mocks/handlers';
import preferenceHandlers from 'datasources/user/preference/mocks/handlers';
import prepaidBalanceServiceHandlers from 'datasources/finance/prepaidBalanceService/mocks/handlers';
import productDetailServiceHandlers from 'datasources/product/details/mocks/handlers';
import productFiltersServiceHandlers from 'datasources/product/filters/mocks/handlers';
import productOffersHandlers from 'datasources/product/offers/mocks/handlers';
import productPromotionsMPDServiceHandlers from 'datasources/product/promotions/mpd/mocks/handlers';
import profileServiceHandlers from 'datasources/marketing/profileService/mocks/handlers';
import resourceHandlers from 'datasources/utility/resource/mocks/handlers';
import roamingRatesHandlers from 'datasources/product/roamingRates/mocks/handlers';
import subscriptionActivationHandlers from 'datasources/line/subscriptionActivationService/mocks/handlers';
import topupHandlers from 'datasources/finance/paymentService/topup/mocks/handlers';
import transactionsListHandlers from 'datasources/line/transactionsService/mocks/handlers';
import troubleshootHandlers from 'datasources/line/assure/connectionTroubleshooter/mocks/handlers';
import userInfoHandlers from 'datasources/user/userInfo/mocks/handlers';
import walletServiceHandlers from 'datasources/finance/walletService/mocks/handlers';

const handlers = [
  ...a2pCustomerHandlers,
  ...accessHandlers,
  ...accountSummaryHandlers,
  ...addressHandlers,
  ...availableNumbers,
  ...airpointsHandlers,
  ...balanceServiceHandlers,
  ...billPaymentHandlers,
  ...cartServicesHandlers,
  ...checkoutServicesHandlers,
  ...connectionDiagnosticsHandlers,
  ...connectionPromiseHandler,
  ...contentServiceHandlers,
  ...digitalBillHandlers,
  ...identityServiceHandlers,
  ...lineConfiguarationHandlers,
  ...lineDetailServiceHandlers,
  ...lineDetailsHandlers,
  ...lineSpendHistoryHandlers,
  ...lineSubscriptionsService,
  ...lineUsageDetailsMocks,
  ...lineUsageHistoryMocks,
  ...lineUsageMocks,
  ...lineUsageTransactionMocks,
  ...networkSettingsHandlers,
  ...orderHandlers,
  ...paymentServiceHandlers,
  ...pointOfAddressHandlers,
  ...pointOfOutageHandlers,
  ...preferenceHandlers,
  ...prepaidBalanceServiceHandlers,
  ...productDetailServiceHandlers,
  ...productFiltersServiceHandlers,
  ...productOffersHandlers,
  ...productPromotionsMPDServiceHandlers,
  ...profileServiceHandlers,
  ...resourceHandlers,
  ...roamingRatesHandlers,
  ...subscriptionActivationHandlers,
  ...topupHandlers,
  ...transactionsListHandlers,
  ...troubleshootHandlers,
  ...userInfoHandlers,
  ...walletServiceHandlers,
];

const server = setupServer(...handlers);
export default server;
