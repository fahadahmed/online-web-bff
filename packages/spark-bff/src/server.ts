import { ApolloServer } from 'apollo-server-fastify';
import {
  ApolloServerPluginLandingPageDisabled,
  ApolloServerPluginLandingPageGraphQLPlayground,
} from 'apollo-server-core';
import { logger } from 'apolloLogger';
import context from 'context';
import dataSources from 'datasources';
import { ApolloLoggerPlugin, ApplicationLogEvent } from 'logging';
import schema from 'schema';
import { formatResponse } from 'utils/formatResponse';
import { isDevelopmentEnvironment } from 'utils/isDevelopmentEnvironment';

process.on('uncaughtException', (err) => {
  logger.error(new ApplicationLogEvent({ message: err.message, context: {} }));
  throw err;
});

const apolloServer = new ApolloServer({
  schema,
  dataSources,
  formatResponse,
  logger,
  context,
  introspection: isDevelopmentEnvironment(),
  plugins: [
    new ApolloLoggerPlugin(
      `BFF Server Started. BUILD_ID: ${process.env.BUILD_BUILDID}`,
    ),
    isDevelopmentEnvironment()
      ? ApolloServerPluginLandingPageGraphQLPlayground()
      : ApolloServerPluginLandingPageDisabled(),
  ],
});

export default apolloServer;
