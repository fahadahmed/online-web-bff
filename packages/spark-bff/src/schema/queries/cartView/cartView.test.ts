import { createTestClient, gql } from 'test/apolloTestUtils';
import { server as mockServer, rest } from 'test/server';
import {
  accountListHandler,
  emptyAccountListHandler,
  guestAccountListHandler,
} from 'datasources/customer/accountsService/mocks/handlers';
import guestEmptyCartMock from 'datasources/shop/cart/mocks/get-cart-guest-empty.mock.json';
import guestPlansCartMock from 'datasources/shop/cart/mocks/get-cart-guest-plans.mock.json';
import sparkIdPlanCartMock from 'datasources/shop/cart/mocks/get-cart-spark-id-plans.mock.json';
import sparkIdPlanCartWithAccountNumberMock from 'datasources/shop/cart/mocks/get-cart-spark-id-plans-with-account-number.mock.json';

const getCartHandler = rest.get(
  'http://testhost/v1/shopping/cart/:cartId',
  async (req, res, ctx) => {
    const cartId = req.params.cartId as string;

    if (cartId === 'success-guest-empty-cart-id') {
      return res(ctx.status(200), ctx.json(guestEmptyCartMock));
    }

    if (cartId === 'success-guest-plans-cart-id') {
      return res(ctx.status(200), ctx.json(guestPlansCartMock));
    }

    if (cartId === 'success-spark-id-plans-cart-id') {
      return res(ctx.status(200), ctx.json(sparkIdPlanCartMock));
    }

    if (cartId === 'success-spark-id-plans-with-account-number-cart-id') {
      return res(
        ctx.status(200),
        ctx.json(sparkIdPlanCartWithAccountNumberMock),
      );
    }

    return res(ctx.status(404));
  },
);

test('cartView responds with null cart and EMPTY_SCREEN for cart error scenario', async () => {
  mockServer.use(getCartHandler, guestAccountListHandler);

  const { query } = createTestClient();

  const res = await query({
    query: gql`
      query CartViewQuery($cartId: String!, $channel: CartChannel!) {
        cartView(cartId: $cartId, channel: $channel) {
          id
          cart {
            id
            cartId
            bundles {
              bundleId
              items {
                itemId
                title
              }
            }
          }
          screen
          mandatoryAction
        }
      }
    `,
    variables: {
      cartId: 'fail-cart-id',
      channel: 'personalshop',
    },
  });
  expect(res.errors).toBeUndefined();
  expect(res.data).toMatchSnapshot();
});

test('cartView responds with cart and EMPTY_SCREEN for empty cart scenario', async () => {
  mockServer.use(getCartHandler, guestAccountListHandler);

  const { query } = createTestClient();

  const res = await query({
    query: gql`
      query CartViewQuery($cartId: String!, $channel: CartChannel!) {
        cartView(cartId: $cartId, channel: $channel) {
          id
          cart {
            id
            cartId
            bundles {
              bundleId
              items {
                itemId
                title
              }
            }
          }
          screen
          mandatoryAction
        }
      }
    `,
    variables: {
      cartId: 'success-guest-empty-cart-id',
      channel: 'personalshop',
    },
  });
  expect(res.errors).toBeUndefined();
  expect(res.data).toMatchSnapshot();
});

test('cartView responds with cart and OVERVIEW_SCREEN for guest scenario', async () => {
  mockServer.use(getCartHandler, guestAccountListHandler);

  const { query } = createTestClient();

  const res = await query({
    query: gql`
      query CartViewQuery($cartId: String!, $channel: CartChannel!) {
        cartView(cartId: $cartId, channel: $channel) {
          id
          cart {
            id
            cartId
            bundles {
              bundleId
              items {
                itemId
                title
              }
            }
          }
          screen
          mandatoryAction
        }
      }
    `,
    variables: {
      cartId: 'success-guest-plans-cart-id',
      channel: 'personalshop',
    },
  });
  expect(res.errors).toBeUndefined();
  expect(res.data).toMatchSnapshot();
});

test('cartView responds with cart, OVERVIEW_SCREEN and CHOOSE_ACCOUNT mandatory action for Spark Id cart without account number scenario', async () => {
  mockServer.use(getCartHandler, accountListHandler);

  const { query } = createTestClient();

  const res = await query({
    query: gql`
      query CartViewQuery($cartId: String!, $channel: CartChannel!) {
        cartView(cartId: $cartId, channel: $channel) {
          id
          cart {
            id
            cartId
            bundles {
              bundleId
              items {
                itemId
                title
              }
            }
          }
          screen
          mandatoryAction
        }
      }
    `,
    variables: {
      cartId: 'success-spark-id-plans-cart-id',
      channel: 'personalshop',
    },
  });
  expect(res.errors).toBeUndefined();
  expect(res.data).toMatchSnapshot();
});

test('cartView responds with cart, OVERVIEW_SCREEN and no mandatory action for Spark Id cart with account number scenario', async () => {
  mockServer.use(getCartHandler, accountListHandler);

  const { query } = createTestClient();

  const res = await query({
    query: gql`
      query CartViewQuery($cartId: String!, $channel: CartChannel!) {
        cartView(cartId: $cartId, channel: $channel) {
          id
          cart {
            id
            cartId
            bundles {
              bundleId
              items {
                itemId
                title
              }
            }
          }
          screen
          mandatoryAction
        }
      }
    `,
    variables: {
      cartId: 'success-spark-id-plans-with-account-number-cart-id',
      channel: 'personalshop',
    },
  });
  expect(res.errors).toBeUndefined();
  expect(res.data).toMatchSnapshot();
});

test('cartView responds with cart and NO_ACCOUNT_SCREEN for Spark Id cart with no accounts scenario', async () => {
  mockServer.use(getCartHandler, emptyAccountListHandler);

  const { query } = createTestClient();

  const res = await query({
    query: gql`
      query CartViewQuery($cartId: String!, $channel: CartChannel!) {
        cartView(cartId: $cartId, channel: $channel) {
          id
          cart {
            id
            cartId
            bundles {
              bundleId
              items {
                itemId
                title
              }
            }
          }
          screen
          mandatoryAction
        }
      }
    `,
    variables: {
      cartId: 'success-spark-id-plans-cart-id',
      channel: 'personalshop',
    },
  });
  expect(res.errors).toBeUndefined();
  expect(res.data).toMatchSnapshot();
});
