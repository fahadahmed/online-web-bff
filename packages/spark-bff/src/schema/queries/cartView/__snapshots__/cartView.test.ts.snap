// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`cartView responds with cart and EMPTY_SCREEN for empty cart scenario 1`] = `
Object {
  "cartView": Object {
    "cart": Object {
      "bundles": Array [],
      "cartId": "success-guest-empty-cart-id",
      "id": "success-guest-empty-cart-id",
    },
    "id": "cart-view",
    "mandatoryAction": null,
    "screen": "EMPTY_SCREEN",
  },
}
`;

exports[`cartView responds with cart and NO_ACCOUNT_SCREEN for Spark Id cart with no accounts scenario 1`] = `
Object {
  "cartView": Object {
    "cart": null,
    "id": "cart-view",
    "mandatoryAction": null,
    "screen": "NO_ACCOUNT_SCREEN",
  },
}
`;

exports[`cartView responds with cart and OVERVIEW_SCREEN for guest scenario 1`] = `
Object {
  "cartView": Object {
    "cart": Object {
      "bundles": Array [
        Object {
          "bundleId": "f4abe282-ac3a-4ba9-9b84-655486784325",
          "items": Array [
            Object {
              "itemId": "f8f3f20a-e100-4de5-b013-6effd77fe792",
              "title": "$79.99 Endless Plan",
            },
            Object {
              "itemId": "7b7360df-f79e-4ee5-9e03-a653d4874315",
              "title": "Spotify Premium",
            },
            Object {
              "itemId": "f68a7ca0-e6e4-4b52-966c-29b6fd63adee",
              "title": "iPhone 13 Pro 128GB - Graphite",
            },
            Object {
              "itemId": "109c861e-e438-4b47-a525-57566f290904",
              "title": "Interest free payment over 24 months",
            },
          ],
        },
        Object {
          "bundleId": "f2b09286-4a6c-4415-8571-c418e8b5ac2b",
          "items": Array [
            Object {
              "itemId": "34ea9285-150a-4f62-a6ff-15cfcbcb0e8f",
              "title": "$20 Prepaid Value Pack",
            },
            Object {
              "itemId": "2ae5a775-dad8-44df-8ec0-88bee05d464b",
              "title": "Spotify Premium",
            },
          ],
        },
        Object {
          "bundleId": "bb7ee884-e02a-43c0-bcca-703ed0ec0c4d",
          "items": Array [
            Object {
              "itemId": "ce7b812f-88a5-47b3-9277-f94bbe095434",
              "title": "Neon",
            },
          ],
        },
        Object {
          "bundleId": "66eb43c7-295b-4d86-ad3d-5e09ce43fc37",
          "items": Array [
            Object {
              "itemId": "2de00c16-9e22-40af-98d2-784b64821baa",
              "title": "$59.99 Endless Plan",
            },
            Object {
              "itemId": "e258d2d9-4bf5-4686-b870-239072ee1f8d",
              "title": "Spotify Premium",
            },
          ],
        },
      ],
      "cartId": "success-guest-plans-cart-id",
      "id": "success-guest-plans-cart-id",
    },
    "id": "cart-view",
    "mandatoryAction": null,
    "screen": "OVERVIEW_SCREEN",
  },
}
`;

exports[`cartView responds with cart, OVERVIEW_SCREEN and CHOOSE_ACCOUNT mandatory action for Spark Id cart without account number scenario 1`] = `
Object {
  "cartView": Object {
    "cart": Object {
      "bundles": Array [
        Object {
          "bundleId": "f4abe282-ac3a-4ba9-9b84-655486784325",
          "items": Array [
            Object {
              "itemId": "f8f3f20a-e100-4de5-b013-6effd77fe792",
              "title": "$79.99 Endless Plan",
            },
            Object {
              "itemId": "7b7360df-f79e-4ee5-9e03-a653d4874315",
              "title": "Spotify Premium",
            },
            Object {
              "itemId": "f68a7ca0-e6e4-4b52-966c-29b6fd63adee",
              "title": "iPhone 13 Pro 128GB - Graphite",
            },
            Object {
              "itemId": "109c861e-e438-4b47-a525-57566f290904",
              "title": "Interest free payment over 24 months",
            },
          ],
        },
        Object {
          "bundleId": "f2b09286-4a6c-4415-8571-c418e8b5ac2b",
          "items": Array [
            Object {
              "itemId": "34ea9285-150a-4f62-a6ff-15cfcbcb0e8f",
              "title": "$20 Prepaid Value Pack",
            },
            Object {
              "itemId": "2ae5a775-dad8-44df-8ec0-88bee05d464b",
              "title": "Spotify Premium",
            },
          ],
        },
        Object {
          "bundleId": "bb7ee884-e02a-43c0-bcca-703ed0ec0c4d",
          "items": Array [
            Object {
              "itemId": "ce7b812f-88a5-47b3-9277-f94bbe095434",
              "title": "Neon",
            },
          ],
        },
        Object {
          "bundleId": "66eb43c7-295b-4d86-ad3d-5e09ce43fc37",
          "items": Array [
            Object {
              "itemId": "2de00c16-9e22-40af-98d2-784b64821baa",
              "title": "$59.99 Endless Plan",
            },
            Object {
              "itemId": "e258d2d9-4bf5-4686-b870-239072ee1f8d",
              "title": "Spotify Premium",
            },
          ],
        },
      ],
      "cartId": "success-spark-id-plans-cart-id",
      "id": "success-spark-id-plans-cart-id",
    },
    "id": "cart-view",
    "mandatoryAction": "CHOOSE_ACCOUNT",
    "screen": "OVERVIEW_SCREEN",
  },
}
`;

exports[`cartView responds with cart, OVERVIEW_SCREEN and no mandatory action for Spark Id cart with account number scenario 1`] = `
Object {
  "cartView": Object {
    "cart": Object {
      "bundles": Array [
        Object {
          "bundleId": "f4abe282-ac3a-4ba9-9b84-655486784325",
          "items": Array [
            Object {
              "itemId": "f8f3f20a-e100-4de5-b013-6effd77fe792",
              "title": "$79.99 Endless Plan",
            },
            Object {
              "itemId": "7b7360df-f79e-4ee5-9e03-a653d4874315",
              "title": "Spotify Premium",
            },
            Object {
              "itemId": "f68a7ca0-e6e4-4b52-966c-29b6fd63adee",
              "title": "iPhone 13 Pro 128GB - Graphite",
            },
            Object {
              "itemId": "109c861e-e438-4b47-a525-57566f290904",
              "title": "Interest free payment over 24 months",
            },
          ],
        },
        Object {
          "bundleId": "f2b09286-4a6c-4415-8571-c418e8b5ac2b",
          "items": Array [
            Object {
              "itemId": "34ea9285-150a-4f62-a6ff-15cfcbcb0e8f",
              "title": "$20 Prepaid Value Pack",
            },
            Object {
              "itemId": "2ae5a775-dad8-44df-8ec0-88bee05d464b",
              "title": "Spotify Premium",
            },
          ],
        },
        Object {
          "bundleId": "bb7ee884-e02a-43c0-bcca-703ed0ec0c4d",
          "items": Array [
            Object {
              "itemId": "ce7b812f-88a5-47b3-9277-f94bbe095434",
              "title": "Neon",
            },
          ],
        },
        Object {
          "bundleId": "66eb43c7-295b-4d86-ad3d-5e09ce43fc37",
          "items": Array [
            Object {
              "itemId": "2de00c16-9e22-40af-98d2-784b64821baa",
              "title": "$59.99 Endless Plan",
            },
            Object {
              "itemId": "e258d2d9-4bf5-4686-b870-239072ee1f8d",
              "title": "Spotify Premium",
            },
          ],
        },
      ],
      "cartId": "success-spark-id-plans-with-account-number-cart-id",
      "id": "success-spark-id-plans-with-account-number-cart-id",
    },
    "id": "cart-view",
    "mandatoryAction": null,
    "screen": "OVERVIEW_SCREEN",
  },
}
`;

exports[`cartView responds with null cart and EMPTY_SCREEN for cart error scenario 1`] = `
Object {
  "cartView": Object {
    "cart": null,
    "id": "cart-view",
    "mandatoryAction": null,
    "screen": "EMPTY_SCREEN",
  },
}
`;
